// _@antv_g@5.0.22@@antv/g/es/index.js
import '../../../_reflect-metadata@0.1.13@reflect-metadata/Reflect.js';
import '../../../_inversify@5.1.1@inversify/lib/inversify.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/core.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/inversify/index.js';
import { GlobalContainer } from '../../../_mana-syringe@0.3.1@mana-syringe/es/container.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/contribution/contribution-protocol.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/register.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/contribution/index.js';
import { containerModule } from './global-module.js';
export { CanvasConfig, LineCap, LineJoin, Shape, TextAlign, TextTransform } from './types.js';
export { Nested, ParenLess, UnitCategory, UnitType, ValueRange } from './css/cssom/types.js';
export { CSSColorValue } from './css/cssom/CSSColorValue.js';
export { CSSKeywordValue } from './css/cssom/CSSKeywordValue.js';
export { CSSMathOperator } from './css/cssom/CSSMathOperator.js';
export { CSSMathInvert } from './css/cssom/CSSMathInvert.js';
export { CSSMathMax } from './css/cssom/CSSMathMax.js';
export { CSSMathMin } from './css/cssom/CSSMathMin.js';
export { CSSMathNegate } from './css/cssom/CSSMathNegate.js';
export { CSSMathProduct } from './css/cssom/CSSMathProduct.js';
export { CSSMathSum } from './css/cssom/CSSMathSum.js';
export { CSSMathValue } from './css/cssom/CSSMathValue.js';
export { CSSMathVariadic, typeCheck } from './css/cssom/CSSMathVariadic.js';
export { CSSNumericValue } from './css/cssom/CSSNumericValue.js';
export { BaseType, CSSNumericValueType, baseTypeToString, unitTypeToBaseType } from './css/cssom/CSSNumericValueType.js';
export { CSSRGB } from './css/cssom/CSSRGB.js';
export { CSSGradientValue, GradientPatternType } from './css/cssom/CSSGradientValue.js';
export { CSSStyleValue, CSSStyleValueType } from './css/cssom/CSSStyleValue.js';
import './css/cssom/CSSTransformValue.js';
import './css/cssom/CSSTranslate.js';
export { CSSUnitValue } from './css/cssom/CSSUnitValue.js';
export { convertAngleUnit, convertPercentUnit, mergeDimensionList, mergeDimensions, parseAngle, parseDimension, parseDimensionArray, parseLength, parseLengthOrPercentage } from './css/parser/dimension.js';
export { Rectangle } from './shapes/Rectangle.js';
export { parseFilter, parseParam } from './css/parser/filter.js';
export { Odeg, Opx, mergeTransforms, parseTransform } from './css/parser/transform.js';
export { CSSPropertyLengthOrPercentage } from './css/properties/CSSPropertyLengthOrPercentage.js';
export { CSSPropertyLocalPosition } from './css/properties/CSSPropertyLocalPosition.js';
export { CSSPropertyOpacity } from './css/properties/CSSPropertyOpacity.js';
export { CSSPropertyColor } from './css/properties/CSSPropertyColor.js';
export { CSSPropertyFilter } from './css/properties/CSSPropertyFilter.js';
export { CSSPropertyLineDash } from './css/properties/CSSPropertyLineDash.js';
export { CSSPropertyShadowBlur } from './css/properties/CSSPropertyShadowBlur.js';
export { CSSPropertyOffsetPath } from './css/properties/CSSPropertyOffsetPath.js';
export { CSSPropertyOffsetDistance } from './css/properties/CSSPropertyOffsetDistance.js';
export { CSSPropertyAnchor } from './css/properties/CSSPropertyAnchor.js';
export { CSSPropertyZIndex } from './css/properties/CSSPropertyZIndex.js';
export { CSSPropertyTransform } from './css/properties/CSSPropertyTransform.js';
export { CSSPropertyTransformOrigin } from './css/properties/CSSPropertyTransformOrigin.js';
export { CSSPropertyPath } from './css/properties/CSSPropertyPath.js';
export { CSSPropertyPoints } from './css/properties/CSSPropertyPoints.js';
export { CSSPropertyClipPath } from './css/properties/CSSPropertyClipPath.js';
export { CSSPropertyText } from './css/properties/CSSPropertyText.js';
export { CSSPropertyTextTransform } from './css/properties/CSSPropertyTextTransform.js';
export { BUILT_IN_PROPERTIES, PropertySyntax, StyleValueRegistry } from './css/StyleValueRegistry.js';
export { LayoutRegistry } from './css/LayoutRegistry.js';
export { ElementEvent } from './dom/interfaces.js';
export { DELEGATION_SPLITTER, EventTarget } from './dom/EventTarget.js';
export { Node } from './dom/Node.js';
export { Element } from './dom/Element.js';
export { Document } from './dom/Document.js';
export { FederatedEvent } from './dom/FederatedEvent.js';
export { FederatedMouseEvent } from './dom/FederatedMouseEvent.js';
export { FederatedPointerEvent } from './dom/FederatedPointerEvent.js';
export { FederatedWheelEvent } from './dom/FederatedWheelEvent.js';
export { Animation } from './dom/Animation.js';
export { AnimationEvent } from './dom/AnimationEvent.js';
export { AnimationEffectTiming } from './dom/AnimationEffectTiming.js';
export { CustomEvent } from './dom/CustomEvent.js';
export { MutationEvent } from './dom/MutationEvent.js';
export { KeyframeEffect, makeTiming, normalizeTimingInput, numericTimingToObject } from './dom/KeyframeEffect.js';
export { AnimationTimeline, compareAnimations } from './dom/AnimationTimeline.js';
export { CustomElementRegistry } from './dom/CustomElementRegistry.js';
export { Camera, CameraEvent, CameraProjectionMode, CameraTrackingMode, CameraType, DefaultCamera } from './camera/Camera.js';
export { DisplayObject } from './display-objects/DisplayObject.js';
import './display-objects/CustomElement.js';
export { Circle } from './display-objects/Circle.js';
export { Ellipse } from './display-objects/Ellipse.js';
export { Group } from './display-objects/Group.js';
export { Image } from './display-objects/Image.js';
export { Line } from './display-objects/Line.js';
export { Path } from './display-objects/Path.js';
export { Polygon } from './display-objects/Polygon.js';
export { Polyline } from './display-objects/Polyline.js';
export { Rect } from './display-objects/Rect.js';
export { Text } from './display-objects/Text.js';
export { HTML } from './display-objects/HTML.js';
export { Geometry } from './components/Geometry.js';
export { Renderable } from './components/Renderable.js';
export { Transform } from './components/Transform.js';
export { Cullable, Strategy } from './components/Cullable.js';
export { Sortable } from './components/Sortable.js';
export { Canvas, CanvasEvent } from './Canvas.js';
export { DisplayObjectPool } from './DisplayObjectPool.js';
export { DefaultSceneGraphService, SceneGraphService, dirtifyToRoot, sortByZIndex } from './services/SceneGraphService.js';
export { ContextService } from './services/ContextService.js';
export { RenderReason, RenderingContext } from './services/RenderingContext.js';
export { RenderingPluginContribution, RenderingService } from './services/RenderingService.js';
export { EventService } from './services/EventService.js';
export { DefaultSceneGraphSelector, SceneGraphSelector, SceneGraphSelectorFactory } from './services/SceneGraphSelector.js';
export { TextService } from './services/TextService.js';
export { OffscreenCanvasCreator } from './services/OffscreenCanvasCreator.js';
export { GeometryAABBUpdater, GeometryUpdaterFactory } from './services/aabb/interfaces.js';
export { CircleUpdater } from './services/aabb/CircleUpdater.js';
export { EllipseUpdater } from './services/aabb/EllipseUpdater.js';
export { RectUpdater } from './services/aabb/RectUpdater.js';
export { TextUpdater } from './services/aabb/TextUpdater.js';
export { LineUpdater } from './services/aabb/LineUpdater.js';
export { PolylineUpdater } from './services/aabb/PolylineUpdater.js';
export { PathUpdater } from './services/aabb/PathUpdater.js';
export { AABB } from './shapes/AABB.js';
import './shapes/BoundingSphere.js';
export { Frustum, Mask } from './shapes/Frustum.js';
export { Plane } from './shapes/Plane.js';
import './shapes/Ray.js';
export { Point } from './shapes/Point.js';
import './utils/custom-easing.js';

GlobalContainer.load(containerModule);
