// _@antv_g@5.0.22@@antv/g/es/types.js
import '../../../_inversify@5.1.1@inversify/lib/inversify.js';
import { Syringe } from '../../../_mana-syringe@0.3.1@mana-syringe/es/core.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/inversify/index.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/container.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/contribution/contribution-protocol.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/register.js';
import '../../../_mana-syringe@0.3.1@mana-syringe/es/contribution/index.js';

var Shape;

(function (Shape) {
  Shape["GROUP"] = "g";
  Shape["CIRCLE"] = "circle";
  Shape["ELLIPSE"] = "ellipse";
  Shape["IMAGE"] = "image";
  Shape["RECT"] = "rect";
  Shape["LINE"] = "line";
  Shape["POLYLINE"] = "polyline";
  Shape["POLYGON"] = "polygon";
  Shape["TEXT"] = "text";
  Shape["PATH"] = "path";
  Shape["HTML"] = "html";
})(Shape || (Shape = {}));

var LineCap;

(function (LineCap) {
  LineCap["BUTT"] = "butt";
  LineCap["ROUND"] = "round";
  LineCap["SQUARE"] = "square";
})(LineCap || (LineCap = {}));

var LineJoin;

(function (LineJoin) {
  LineJoin["BEVEL"] = "bevel";
  LineJoin["ROUND"] = "round";
  LineJoin["MITER"] = "miter";
})(LineJoin || (LineJoin = {}));

var TextTransform;

(function (TextTransform) {
  TextTransform["CAPITALIZE"] = "capitalize";
  TextTransform["UPPERCASE"] = "uppercase";
  TextTransform["LOWERCASE"] = "lowercase";
  TextTransform["NONE"] = "none";
})(TextTransform || (TextTransform = {}));

var TextAlign;

(function (TextAlign) {
  TextAlign["START"] = "start";
  TextAlign["CENTER"] = "center";
  TextAlign["END"] = "end";
  TextAlign["LEFT"] = "left";
  TextAlign["RIGHT"] = "right";
})(TextAlign || (TextAlign = {}));

var CanvasConfig = Syringe.defineToken('CanvasConfig');

export { CanvasConfig, LineCap, LineJoin, Shape, TextAlign, TextTransform };
