// inversify/lib/planning/request.js
import { __exports as request } from '../../../../_virtual/request.js';
export { __exports as default } from '../../../../_virtual/request.js';
import '../utils/id.js';
import { __exports as id } from '../../../../_virtual/id.js';

Object.defineProperty(request, "__esModule", { value: true });
request.Request = void 0;
var id_1 = id;
var Request = (function () {
    function Request(serviceIdentifier, parentContext, parentRequest, bindings, target) {
        this.id = id_1.id();
        this.serviceIdentifier = serviceIdentifier;
        this.parentContext = parentContext;
        this.parentRequest = parentRequest;
        this.target = target;
        this.childRequests = [];
        this.bindings = (Array.isArray(bindings) ? bindings : [bindings]);
        this.requestScope = parentRequest === null
            ? new Map()
            : null;
    }
    Request.prototype.addChildRequest = function (serviceIdentifier, bindings, target) {
        var child = new Request(serviceIdentifier, this.parentContext, this, bindings, target);
        this.childRequests.push(child);
        return child;
    };
    return Request;
}());
request.Request = Request;
