// @antv/g/es/services/aabb/CircleUpdater.js
import { __decorate } from '../../../../../tslib/tslib.es6.js';
import { singleton } from '../../../../../mana-syringe/es/decorator.js';
import '../../../../../mana-syringe/es/container.js';
import '../../../../../mana-syringe/es/core.js';
import '../../../../../mana-syringe/es/contribution/contribution-protocol.js';
import '../../../../../inversify/lib/inversify.js';
import '../../../../../mana-syringe/es/register.js';
import '../../../../../mana-syringe/es/contribution/index.js';
import { GeometryAABBUpdater } from './interfaces.js';
import { Shape } from '../../types.js';

var CircleUpdater =
/** @class */
function () {
  function CircleUpdater() {}

  CircleUpdater.prototype.update = function (parsedStyle, object) {
    var r = parsedStyle.r,
        x = parsedStyle.x,
        y = parsedStyle.y;
    var width = r.value * 2;
    var height = r.value * 2;
    return {
      width: width,
      height: height,
      x: x.value || 0,
      y: y.value || 0
    };
  };

  CircleUpdater = __decorate([singleton({
    token: {
      token: GeometryAABBUpdater,
      named: Shape.CIRCLE
    }
  })], CircleUpdater);
  return CircleUpdater;
}();

export { CircleUpdater };
