// @antv/g/es/services/aabb/EllipseUpdater.js
import { __decorate } from '../../../../../tslib/tslib.es6.js';
import { singleton } from '../../../../../mana-syringe/es/decorator.js';
import '../../../../../mana-syringe/es/container.js';
import '../../../../../mana-syringe/es/core.js';
import '../../../../../mana-syringe/es/contribution/contribution-protocol.js';
import '../../../../../inversify/lib/inversify.js';
import '../../../../../mana-syringe/es/register.js';
import '../../../../../mana-syringe/es/contribution/index.js';
import { GeometryAABBUpdater } from './interfaces.js';
import { Shape } from '../../types.js';

var EllipseUpdater =
/** @class */
function () {
  function EllipseUpdater() {}

  EllipseUpdater.prototype.update = function (parsedStyle, object) {
    var rx = parsedStyle.rx,
        ry = parsedStyle.ry,
        x = parsedStyle.x,
        y = parsedStyle.y;
    var width = rx.value * 2;
    var height = ry.value * 2;
    return {
      width: width,
      height: height,
      x: x.value || 0,
      y: y.value || 0
    };
  };

  EllipseUpdater = __decorate([singleton({
    token: {
      token: GeometryAABBUpdater,
      named: Shape.ELLIPSE
    }
  })], EllipseUpdater);
  return EllipseUpdater;
}();

export { EllipseUpdater };
