// @antv/g/es/services/aabb/LineUpdater.js
import { __decorate } from '../../../../../tslib/tslib.es6.js';
import { singleton } from '../../../../../mana-syringe/es/decorator.js';
import '../../../../../mana-syringe/es/container.js';
import '../../../../../mana-syringe/es/core.js';
import '../../../../../mana-syringe/es/contribution/contribution-protocol.js';
import '../../../../../inversify/lib/inversify.js';
import '../../../../../mana-syringe/es/register.js';
import '../../../../../mana-syringe/es/contribution/index.js';
import { GeometryAABBUpdater } from './interfaces.js';
import { Shape } from '../../types.js';

var LineUpdater =
/** @class */
function () {
  function LineUpdater() {}

  LineUpdater.prototype.update = function (parsedStyle) {
    var x1 = parsedStyle.x1,
        y1 = parsedStyle.y1,
        x2 = parsedStyle.x2,
        y2 = parsedStyle.y2;
    var minX = Math.min(x1.value, x2.value);
    var maxX = Math.max(x1.value, x2.value);
    var minY = Math.min(y1.value, y2.value);
    var maxY = Math.max(y1.value, y2.value);
    var width = maxX - minX;
    var height = maxY - minY;
    return {
      width: width,
      height: height,
      x: minX,
      y: minY
    };
  };

  LineUpdater = __decorate([singleton({
    token: {
      token: GeometryAABBUpdater,
      named: Shape.LINE
    }
  })], LineUpdater);
  return LineUpdater;
}();

export { LineUpdater };
