// _inversify@5.1.1@inversify/lib/inversify.js
import { __exports as inversify } from '../../../../../_virtual/inversify.js';
import './constants/metadata_keys.js';
import './container/container.js';
import './constants/literal_types.js';
import './container/container_module.js';
import './annotation/injectable.js';
import './annotation/tagged.js';
import './annotation/named.js';
import './annotation/inject.js';
import './annotation/optional.js';
import './annotation/unmanaged.js';
import './annotation/multi_inject.js';
import './annotation/target_name.js';
import './annotation/post_construct.js';
import './planning/metadata_reader.js';
import './utils/id.js';
import './annotation/decorator_utils.js';
import './syntax/constraint_helpers.js';
import './utils/serialization.js';
import './utils/binding_utils.js';
import { __exports as metadata_keys } from '../../../../../_virtual/metadata_keys.js';
import { __exports as container } from '../../../../../_virtual/container.js';
import { __exports as literal_types } from '../../../../../_virtual/literal_types.js';
import { __exports as container_module } from '../../../../../_virtual/container_module.js';
import { __exports as injectable } from '../../../../../_virtual/injectable.js';
import { __exports as tagged } from '../../../../../_virtual/tagged.js';
import { __exports as named } from '../../../../../_virtual/named.js';
import { __exports as inject } from '../../../../../_virtual/inject.js';
import { __exports as optional } from '../../../../../_virtual/optional.js';
import { __exports as unmanaged } from '../../../../../_virtual/unmanaged.js';
import { __exports as multi_inject } from '../../../../../_virtual/multi_inject.js';
import { __exports as target_name } from '../../../../../_virtual/target_name.js';
import { __exports as post_construct } from '../../../../../_virtual/post_construct.js';
import { __exports as metadata_reader } from '../../../../../_virtual/metadata_reader.js';
import { __exports as id } from '../../../../../_virtual/id.js';
import { __exports as decorator_utils } from '../../../../../_virtual/decorator_utils.js';
import { __exports as constraint_helpers } from '../../../../../_virtual/constraint_helpers.js';
import { __exports as serialization } from '../../../../../_virtual/serialization.js';
import { __exports as binding_utils } from '../../../../../_virtual/binding_utils.js';

(function (exports) {
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiBindToService = exports.getServiceIdentifierAsString = exports.typeConstraint = exports.namedConstraint = exports.taggedConstraint = exports.traverseAncerstors = exports.decorate = exports.id = exports.MetadataReader = exports.postConstruct = exports.targetName = exports.multiInject = exports.unmanaged = exports.optional = exports.LazyServiceIdentifer = exports.inject = exports.named = exports.tagged = exports.injectable = exports.ContainerModule = exports.AsyncContainerModule = exports.TargetTypeEnum = exports.BindingTypeEnum = exports.BindingScopeEnum = exports.Container = exports.METADATA_KEY = void 0;
var keys = metadata_keys;
exports.METADATA_KEY = keys;
var container_1 = container;
Object.defineProperty(exports, "Container", { enumerable: true, get: function () { return container_1.Container; } });
var literal_types_1 = literal_types;
Object.defineProperty(exports, "BindingScopeEnum", { enumerable: true, get: function () { return literal_types_1.BindingScopeEnum; } });
Object.defineProperty(exports, "BindingTypeEnum", { enumerable: true, get: function () { return literal_types_1.BindingTypeEnum; } });
Object.defineProperty(exports, "TargetTypeEnum", { enumerable: true, get: function () { return literal_types_1.TargetTypeEnum; } });
var container_module_1 = container_module;
Object.defineProperty(exports, "AsyncContainerModule", { enumerable: true, get: function () { return container_module_1.AsyncContainerModule; } });
Object.defineProperty(exports, "ContainerModule", { enumerable: true, get: function () { return container_module_1.ContainerModule; } });
var injectable_1 = injectable;
Object.defineProperty(exports, "injectable", { enumerable: true, get: function () { return injectable_1.injectable; } });
var tagged_1 = tagged;
Object.defineProperty(exports, "tagged", { enumerable: true, get: function () { return tagged_1.tagged; } });
var named_1 = named;
Object.defineProperty(exports, "named", { enumerable: true, get: function () { return named_1.named; } });
var inject_1 = inject;
Object.defineProperty(exports, "inject", { enumerable: true, get: function () { return inject_1.inject; } });
Object.defineProperty(exports, "LazyServiceIdentifer", { enumerable: true, get: function () { return inject_1.LazyServiceIdentifer; } });
var optional_1 = optional;
Object.defineProperty(exports, "optional", { enumerable: true, get: function () { return optional_1.optional; } });
var unmanaged_1 = unmanaged;
Object.defineProperty(exports, "unmanaged", { enumerable: true, get: function () { return unmanaged_1.unmanaged; } });
var multi_inject_1 = multi_inject;
Object.defineProperty(exports, "multiInject", { enumerable: true, get: function () { return multi_inject_1.multiInject; } });
var target_name_1 = target_name;
Object.defineProperty(exports, "targetName", { enumerable: true, get: function () { return target_name_1.targetName; } });
var post_construct_1 = post_construct;
Object.defineProperty(exports, "postConstruct", { enumerable: true, get: function () { return post_construct_1.postConstruct; } });
var metadata_reader_1 = metadata_reader;
Object.defineProperty(exports, "MetadataReader", { enumerable: true, get: function () { return metadata_reader_1.MetadataReader; } });
var id_1 = id;
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return id_1.id; } });
var decorator_utils_1 = decorator_utils;
Object.defineProperty(exports, "decorate", { enumerable: true, get: function () { return decorator_utils_1.decorate; } });
var constraint_helpers_1 = constraint_helpers;
Object.defineProperty(exports, "traverseAncerstors", { enumerable: true, get: function () { return constraint_helpers_1.traverseAncerstors; } });
Object.defineProperty(exports, "taggedConstraint", { enumerable: true, get: function () { return constraint_helpers_1.taggedConstraint; } });
Object.defineProperty(exports, "namedConstraint", { enumerable: true, get: function () { return constraint_helpers_1.namedConstraint; } });
Object.defineProperty(exports, "typeConstraint", { enumerable: true, get: function () { return constraint_helpers_1.typeConstraint; } });
var serialization_1 = serialization;
Object.defineProperty(exports, "getServiceIdentifierAsString", { enumerable: true, get: function () { return serialization_1.getServiceIdentifierAsString; } });
var binding_utils_1 = binding_utils;
Object.defineProperty(exports, "multiBindToService", { enumerable: true, get: function () { return binding_utils_1.multiBindToService; } });

}(inversify));
